### PR TITLE
split: suffix length refactor

### PR DESCRIPTION
**Pull request by zhitkoff**
Refactoring suffix length processing, distinguishing between dynamic auto-widening vs. pre-calculated auto-width and addressing an edge case of passing `0` as a value for suffix length in command line option, 
i.e. `split -a0 file` 
or `split --suffix-length=0 file`

### CODE REVIEW

**Comment by zhitkoff** [comment]
@sylvestre would you mind reviewing this one?

**Review by tertsdiepraam — COMMENTED** [review]
Cool! Nice tests! I have a few small suggestions, but looks good overall, especially given all the edge case we have to take into account.

File: src/uu/split/src/split.rs [hunk 1]
```diff
@@ -701,36 +722,74 @@ fn suffix_from(
...
-    let suffix_length: usize = suffix_length_str
+    let mut suffix_length: usize = suffix_length_str
         .parse()
         .map_err(|_| SettingsError::SuffixNotParsable(suffix_length_str.to_string()))?;
 
-    // Override suffix_auto_widening if suffix length value came from command line
-    // and not from default value
-    if matches.value_source(OPT_SUFFIX_LENGTH) == Some(ValueSource::CommandLine) {
+    // Disable dynamic auto-widening if suffix length was specified in command line with value > 0
+    if matches.value_source(OPT_SUFFIX_LENGTH) == Some(ValueSource::CommandLine)
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs:755]: so, for this option we need to account for the following scenarios as they all have different outcomes, especially in combination with other options (like `--numeric-suffixes`, `--hex-suffixes` and `--number`):
1) value was not provided in cmd line
2) value was provided in cmd line and it is > 0
3) value was provided in cmd line and it is == 0

and these need to be done at least in two different places.

If we drop default value on this options, we would still need to distinguish between these 3 scenarios in all applicable combinations. It would just be different checks and possibly parsing to `usize` more then once or introducing additional temp variables, since we cannot do all of these checks in one place - due to how it works in combination with other options. 

I will try a few different ways and see if it can be simplified.

File: src/uu/split/src/split.rs [hunk 2]
```diff
@@ -668,17 +668,38 @@ impl Strategy {
...
+///   as in `split --numeric-suffixes=100 --number=r/100 file`
+///   OR when suffix length N is specified, except for N=0 (see edge cases below)
+///   `-a N` or `--suffix-length=N`
+///
+/// Edge case:
+/// - If suffix length is specified as 0 AND `-n`/`--number` option used specifying number of files:
+///   set auto widening OFF AND auto pre-calculate required suffix length based on number of files needed
+/// - If suffix length is specified as 0 in any other situation
+///   keep auto widening ON and suffix length to default value
+///
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs:702]: yep, I considered it. The best way is to separate suffix (or whole `Settings`) and `Strategy` into their own modules/files though, otherwise it would be cumbersome to pass this new `Suffix` type to `FilenameIterator` in filenames.rs. And if we keep the current way suffix values are passed to it, then there is no much simplification - just in calling `suffix_from`.
Would you be OK with that approach?

File: src/uu/split/src/split.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs]: pushed new version that should address this one as well

**Comment by zhitkoff** [comment]
@tertsdiepraam with latest changes, could you please let me know if you have other comments/suggestions?

File: src/uu/split/src/filenames.rs [hunk 3]
```diff
@@ -54,21 +67,200 @@ impl SuffixType {
...
+        // Set it to default value
+        if is_length_cmd_opt && length == 0 {
+            length = default_length;
+        }
+
+        let additional = matches
+            .get_one::<String>(OPT_ADDITIONAL_SUFFIX)
+            .unwrap()
+            .to_string();
+        if additional.contains('/') {
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/filenames.rs]: makes sense, pushed an update

File: src/uu/split/src/filenames.rs [hunk 4]
```diff
@@ -54,21 +67,200 @@ impl SuffixType {
...
+        // Set it to default value
+        if is_length_cmd_opt && length == 0 {
+            length = default_length;
+        }
+
+        let additional = matches
+            .get_one::<String>(OPT_ADDITIONAL_SUFFIX)
+            .unwrap()
+            .to_string();
+        if additional.contains('/') || additional.contains('\\') {
```
Comment (tertsdiepraam) [inline src/uu/split/src/filenames.rs:233]: I think the check should probably be done with this function https://doc.rust-lang.org/std/path/fn.is_separator.html to be entirely correct, because only `/` should work on Unix and both `\` and `/` should work on Windows (and there might be other exceptions for other platforms).

It would probably look something like this:

```suggestion
        if additional.chars().any(std::path::is_separator) {
```

Mention me again when it's done and we'll merge this :)

**Review by tertsdiepraam — APPROVED** [review]
Great, thanks! Can be merged once the CI is green!

**Comment by zhitkoff** [comment]
@tertsdiepraam looks like it is good to merge. The GNU test failures are unrelated - something is not right with `tail`

**Comment by cakebaker** [comment]
Thanks!
